File: 74.Search a 2D Matrix/74.Search_a_2D_Matrix.java
class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        int rows = matrix.length;
        int columns = matrix[0].length;
        int start = 0;
        int end = rows*columns-1;
        int middle = start + (end-start)/2;
        int temp;
        while(start <= end)
        {
            int element = matrix[middle/columns][middle%columns];
            if(element == target) return true;
            if(element < target) start = middle+1;
            else end = middle-1;
            middle = start + (end-start)/2;
        }
        return false;
    }
}

File: 74.Search-a-2D-Matrix/74.Search_a_2D_Matrix.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int n = matrix.size();
        int m = matrix[0].size();
        int start = 0;
        int end = n*m-1;
        while(start <= end)
        {
            int mid = start + (end-start)/2;
            int x = matrix[mid/m][mid%m];
            if(x > target)
            {
                end = mid - 1;
            }
            else if(x < target)
            {
                start = mid + 1;
            }
            else
            {
                return true;
            }
        }
        return false;
    }
};